core/src/main/java/com/linecorp/armeria/common/LeakTracingRequestContextStorage.java

```diff
@@ -0,0 +1,110 @@
+    private final Sampler<Class<? extends Throwable>> sampler;
+
+    /**
+     * Creates a new instance.
+     * @param delegate the underlying {@link RequestContextStorage} that stores {@link RequestContext}
+     */
+    public LeakTracingRequestContextStorage(RequestContextStorage delegate) {
+        this(delegate, Flags.verboseExceptionSampler());
+    }
+
+    /**
+     * Creates a new instance.
+     * @param delegate the underlying {@link RequestContextStorage} that stores {@link RequestContext}
+     * @param sampler the {@link Sampler} that determines whether to retain the stacktrace of the context leaks
+     */
+    public LeakTracingRequestContextStorage(RequestContextStorage delegate,
+                                            Sampler<Class<? extends Throwable>> sampler) {
+        this.delegate = delegate;
+        pendingRequestCtx = new FastThreadLocal<>();
+        this.sampler = sampler;
+    }
+
+    @Nullable
+    @Override
+    public <T extends RequestContext> T push(RequestContext toPush) {
+        requireNonNull(toPush, "toPush");
+
+        final RequestContext prevContext = delegate.currentOrNull();
+        if (prevContext != null) {
+            if (prevContext == toPush) {
+                // Re-entrance
+            } else if (toPush instanceof ServiceRequestContext &&
+                       prevContext.root() == toPush) {
+                // The delegate has the ServiceRequestContext whose root() is toPush
+            } else if (toPush instanceof ClientRequestContext &&
+                       prevContext.root() == toPush.root()) {
+                // The delegate has the ClientRequestContext whose root() is the same toPush.root()
+            } else {
+                throw pendingRequestCtx.get();
+            }
+        }
+
+        if (sampler.isSampled(PendingRequestContextStackTrace.class)) {
+            pendingRequestCtx.set(new PendingRequestContextStackTrace(toPush, true));
+        } else {
+            pendingRequestCtx.set(new PendingRequestContextStackTrace(toPush, false));
+        }
+
+        return delegate.push(toPush);
+    }
+
+    @Override
+    public void pop(RequestContext current, @Nullable RequestContext toRestore) {
+        delegate.pop(current, toRestore);
+        pendingRequestCtx.remove();
+    }
+
+    @Nullable
+    @Override
+    public <T extends RequestContext> T currentOrNull() {
+        return delegate.currentOrNull();
+    }
+
+    static class PendingRequestContextStackTrace extends RuntimeException {
+
+        private static final long serialVersionUID = -689451606253441556L;
+
+        PendingRequestContextStackTrace(RequestContext context, boolean isSample) {
+            super("At thread [" + currentThread().getName() + "], previous RequestContext didn't popped : " +
+                  context + (isSample ? ", It is pushed at the following stacktrace" : ""), null,
+                  true, isSample);
```

Comment:
I'm not 100% sure but would be efficient if we build the message lazy when `pendingRequestCtx.get()` is called?
```java
PendingRequestContextStackTrace pendingRequestContext = pendingRequestCtx.get();
if (pendingRequestContext != null) {
    throw new IllegalStateException("At thread [" + pendingRequestContext.thread() + ...);
}

static class PendingRequestContextStackTrace extends RuntimeException {
      private final Thread thread;
      private final RequestContext ctx;

      ....
}
```

Reply:
> Therefore, I think the performance might not be a big deal for this feature.

Right. I think this feature would be super useful to debug some trick issues that are hard to reproduce such as #3945
But it is also great we can provide optimizations because some features are not easy to detect in test environments.
They want to turn on this in a canary deployment.

Reply:
On second thought, regardless of performance, it would be helpful to provide both the previous and current stack traces for easier debugging.

Reply:
Ok got it, I agreed.